File: santa-tracker/src/main/java/com/google/android/apps/santatracker/games/cityquiz/CityQuizActivity.java
/*
 * Copyright (C) 2016 Google Inc. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.apps.santatracker.games.cityquiz;

import android.content.Context;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.apps.santatracker.R;
import com.google.android.apps.santatracker.games.EndOfGameView;
import com.google.android.apps.santatracker.games.common.PlayGamesActivity;
import com.google.android.apps.santatracker.launch.StartupActivity;
import com.google.android.apps.santatracker.util.FontHelper;
import com.google.android.apps.santatracker.util.MapHelper;
import com.google.android.apps.santatracker.util.MeasurementManager;
import com.google.android.apps.santatracker.util.SantaLog;
import com.google.android.apps.santatracker.util.VectorUtil;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.MapStyleOptions;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.storage.FirebaseStorage;

import java.lang.ref.WeakReference;

/**
 * Main container for the City Quiz game.
 */
public class CityQuizActivity extends PlayGamesActivity implements
        OnMapReadyCallback,
        GoogleMap.OnMarkerClickListener,
        ViewTreeObserver.OnGlobalLayoutListener {

    private static final String TAG = CityQuizActivity.class.getSimpleName();
    private static final int NEXT_ROUND_DELAY = 3500;
    public static final int CORRECT_MARKER = 0;
    public static final int FIRST_FAKE_MARKER = 1;
    public static final int SECOND_FAKE_MARKER = 2;
    public static final int IMAGE_LOAD_TIMEOUT_MILLIS = 5000; // 5 seconds
    public static final String CITY_QUIZ_ROUND_COUNT_CONFIG_KEY = "CityQuizRoundCount";

    private CityQuizGame mCityQuizGame;

    private ImageView mCityImageView;
    private ProgressBar mCityImageProgressBar;
    private ImageView mCloudOffImageView;
    private CityImageGlideListener mCityImageGlideListener;
    private TextView mCityImageAuthorTextView;
    private SupportMapFragment mSupportMapFragment;
    private GoogleMap mGoogleMap;
    private boolean mMapLaidOut;
    private boolean mMapReady;
    private boolean mInitialRoundLoaded;
    private TextView mRoundCountTextView;
    private TextView mPointsTextView;
    private View mMapScrim;

    private Handler mHandler;

    private FirebaseAnalytics mAnalytics;

    public CityQuizActivity() {
        super(R.layout.activity_city_quiz, StartupActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // [ANALYTICS]
        mAnalytics = FirebaseAnalytics.getInstance(this);
        MeasurementManager.recordScreenView(mAnalytics,
                getString(R.string.analytics_screen_city_quiz));

        mCityImageView = (ImageView) findViewById(R.id.cityQuizImageView);
        mCityImageProgressBar = (ProgressBar) findViewById(R.id.cityImageProgressBar);
        mCloudOffImageView = (ImageView) findViewById(R.id.cloudOffImageView);
        mCityImageAuthorTextView = (TextView) findViewById(R.id.cityImageAuthorTextView);
        mMapScrim = findViewById(R.id.map_scrim);

        mCityImageGlideListener = new CityImageGlideListener(this);

        // Clicking the "offline" image will attempt a reload
        mCloudOffImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CityQuizRound round = mCityQuizGame.getCurrentRound();
                if (round != null && round.getCity() != null) {
                    loadCityImage(round.getCity().getImageName(), round.getCity().getImageAuthor());
                }
            }
        });

        // Map scrim prevents clicking on map
        mMapScrim.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // No-op, eat the click
            }
        });

        // Santa-fonts
        FontHelper.makeLobster((TextView) findViewById(R.id.title_city_quiz));
        FontHelper.makeLobster((TextView) findViewById(R.id.text_what_city));

        mSupportMapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);

        mSupportMapFragment.getView().getViewTreeObserver().addOnGlobalLayoutListener(this);
        mSupportMapFragment.getMapAsync(this);

        long roundCount = FirebaseRemoteConfig.getInstance().getLong(CITY_QUIZ_ROUND_COUNT_CONFIG_KEY);
        mCityQuizGame = new CityQuizGame(this, (int) roundCount);

        mRoundCountTextView = (TextView) findViewById(R.id.roundCountTextView);
        mPointsTextView = (TextView) findViewById(R.id.pointsTextView);

        mHandler = new Handler();
    }

    @Override
    public String getGameId() {
        return getResources().getString(R.string.city_quiz_game_id);
    }

    @Override
    public String getGameTitle() {
        return getResources().getString(R.string.cityquiz);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMapReady = true;
        mGoogleMap = googleMap;
        mGoogleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(this, R.raw.map_style));
        mGoogleMap.setOnMarkerClickListener(this);
        loadInitialRound();
    }

    @Override
    public void onGlobalLayout() {
        mMapLaidOut = true;
        loadInitialRound();
    }

    /**
     * Load the initial round only if the map is ready and laid out and has not been already loaded.
     */
    private synchronized void loadInitialRound() {
        if (mMapLaidOut && mMapReady && !mInitialRoundLoaded) {
            loadRound();
            mInitialRoundLoaded = true;
        }
    }

    /**
     * Load the current round of the game for user interaction.
     */
    public void loadRound() {
        updateScore();

        // Get next city in game.
        final City city = mCityQuizGame.getCurrentRound().getCity();

        // Load city image
        loadCityImage(city.getImageName(), city.getImageAuthor());

        // Set up city markers
        mGoogleMap.clear();

        Log.d(TAG, "Moving to " + city.getName());
        mGoogleMap.animateCamera(CameraUpdateFactory.newLatLngBounds(city.getBounds(),
                MapHelper.getMapPadding(mSupportMapFragment)));

        // Add markers and set appropriate tags.
        Marker locationMarker = mGoogleMap.addMarker(new MarkerOptions()
                .position(city.getCorrectLocation())
                .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)));
        locationMarker.setTag(CORRECT_MARKER);
        Marker firstFakeLocationMarker = mGoogleMap.addMarker(new MarkerOptions()
                .position(city.getIncorrectLocation(City.FIRST_FAKE_INDEX))
                .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)));
        firstFakeLocationMarker.setTag(FIRST_FAKE_MARKER);
        Marker secondFakeLocationMarker = mGoogleMap.addMarker(new MarkerOptions()
                .position(city.getIncorrectLocation(City.SECOND_FAKE_INDEX))
                .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)));
        secondFakeLocationMarker.setTag(SECOND_FAKE_MARKER);
    }

    /**
     * Load the image matching the given image name into the city ImageView.
     *
     * @param imageName Name used to retrieve the image from Firebase Storage.
     * @param imageAuthor Name used to give attribution for the image.
     */
    private void loadCityImage(final String imageName, final String imageAuthor) {
        // Clear current image
        Glide.with(CityQuizActivity.this)
                .load(R.color.cityQuizPrimaryGreenDark)
                .crossFade()
                .into(mCityImageView);

        mCityImageProgressBar.setVisibility(View.VISIBLE);
        showImageAuthor(false);
        showOnlineUI();

        // Load new image
        final Task<Uri> task = FirebaseStorage.getInstance().getReference().child(imageName).getDownloadUrl();
        task.addOnCompleteListener(this, new OnCompleteListener<Uri>() {
                    @Override
                    public void onComplete(@NonNull Task<Uri> task) {
                        mCityImageProgressBar.setVisibility(View.GONE);

                        if (task.isSuccessful()) {
                            Uri uri = task.getResult();
                            Log.d(TAG, "Image uri: " + uri.toString());

                            final Context context = CityQuizActivity.this;
                            Glide.with(context)
                                    .load(uri)
                                    .crossFade()
                                    .listener(mCityImageGlideListener)
                                    .into(mCityImageView);

                            mCityImageAuthorTextView.setText(getResources().getString(R.string.photo_by, imageAuthor));
                            showImageAuthor(true);
                        } else {
                            Log.e(TAG, "Unable to get image URI from Firebase Storage. " + task.getException(),
                                    task.getException());
                            showOfflineUI();
                        }
                    }
                });

        // After timeout check if image URL has been retrieved, if not update UI.
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (!task.isComplete()) {
                    // Firebase was did not complete image URL retrieval.
                    mCityImageProgressBar.setVisibility(View.GONE);
                    showOfflineUI();
                }
            }
        }, IMAGE_LOAD_TIMEOUT_MILLIS);
    }

    private void showOfflineUI() {
        mCloudOffImageView.setVisibility(View.VISIBLE);
        mMapScrim.setVisibility(View.VISIBLE);
    }

    private void showOnlineUI() {
        mCloudOffImageView.setVisibility(View.GONE);
        mMapScrim.setVisibility(View.GONE);
    }

    private void updateScore() {
        // Update game information, round count and score.
        mRoundCountTextView.setText(getString(R.string.round_count_fmt, (mCityQuizGame.getCurrentRoundCount() + 1),
                mCityQuizGame.getTotalRoundCount()));
        mPointsTextView.setText(getString(R.string.game_score_fmt, mCityQuizGame.calculateScore()));
    }

    private void showImageAuthor(boolean visible) {
        if (visible) {
            mCityImageAuthorTextView.setVisibility(TextView.VISIBLE);
        } else {
            mCityImageAuthorTextView.setVisibility(TextView.INVISIBLE);
        }
    }

    @Override
    public boolean onMarkerClick(final Marker marker) {
        // Check if the round is already solved, if so ignore the marker click.
        if (mCityQuizGame.isFinished() || mCityQuizGame.getCurrentRound().isSolved()) {
            return true;
        }
        // Identify which marker was tapped and update the round status.
        int tag = (int)marker.getTag();
        mCityQuizGame.getCurrentRound().updateLocationStatus(tag, true);

        // Check if user tapped on the correct marker and if move to the next round.
        if (tag == 0) {
            marker.setIcon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_green));
            updateScore();

            // [ANALYTICS]
            int numIncorrectAttempts = 5 - mCityQuizGame.getCurrentRound().calculateRoundScore();
            MeasurementManager.recordCorrectCitySelected(mAnalytics,
                    mCityQuizGame.getCurrentRound().getCity().getImageName(),
                    numIncorrectAttempts);

            mGoogleMap.animateCamera(CameraUpdateFactory.newLatLng(mCityQuizGame.getCurrentRound()
                    .getCity().getCorrectLocation()), new GoogleMap.CancelableCallback() {
                @Override
                public void onFinish() {
                    showCityInfo(marker);
                }

                @Override
                public void onCancel() {
                    showCityInfo(marker);
                }
            });

            // Wait a while before moving to next round or end of game.
            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    mCityQuizGame.moveToNextRound();
                    // Check if the last round has completed.
                    if (!mCityQuizGame.isFinished()) {
                        loadRound();
                    } else {
                        // or exit. For now I will clear the markers.
                        goToGameSummary();
                        mPointsTextView.setText(getString(R.string.game_score_fmt, mCityQuizGame.calculateScore()));
                    }
                }
            }, NEXT_ROUND_DELAY);
        } else {
            // [ANALYTICS]
            MeasurementManager.recordIncorrectCitySelected(mAnalytics,
                    mCityQuizGame.getCurrentRound().getCity().getImageName());

            marker.setIcon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_red));
        }

        return true;
    }

    private void showCityInfo(Marker marker) {
        marker.setTitle(mCityQuizGame.getCurrentRound().getCity().getName());
        marker.showInfoWindow();
    }

    private void goToGameSummary() {
        // Show the end-game view
        EndOfGameView gameView = (EndOfGameView) findViewById(R.id.view_end_game);
        gameView.initialize(mCityQuizGame.calculateScore(),
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        // Restart this activity
                        recreate();
                    }

                }, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        // Back to the village
                        finish();
                    }
                });

        // Show end game view over everything
        gameView.setVisibility(View.VISIBLE);
        if (Build.VERSION.SDK_INT >= 21) {
            gameView.setZ(1000f);
        }
    }

    // Listener to handle Glide action completion.
    private static class CityImageGlideListener implements RequestListener {

        private WeakReference<CityQuizActivity> weakActivity;

        private CityImageGlideListener(CityQuizActivity activity) {
            weakActivity = new WeakReference<>(activity);
        }

        @Override
        public boolean onException(Exception e, Object model, Target target, boolean isFirstResource) {
            // Glide failed to load image.
            if (weakActivity.get() != null) {
                SantaLog.e(TAG, "Glide unable to load city image.");
                weakActivity.get().showOfflineUI();
            }
            return false;
        }

        @Override
        public boolean onResourceReady(Object resource, Object model, Target target,
                                       boolean isFromMemoryCache, boolean isFirstResource) {
            if (weakActivity.get() != null) {
                // Glide loaded image, hide "cloud off" ImageView and show author TextView.
                weakActivity.get().showOnlineUI();
                weakActivity.get().showImageAuthor(true);
            }
            return false;
        }
    }
}

File: cityquiz/src/main/java/com/google/android/apps/santatracker/cityquiz/CityQuizActivity.kt
/*
 * Copyright 2019. Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.santatracker.cityquiz

import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.os.postDelayed
import androidx.core.view.doOnLayout
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.google.android.apps.playgames.common.PlayGamesActivity
import com.google.android.apps.santatracker.games.EndOfGameView
import com.google.android.apps.santatracker.util.MapHelper
import com.google.android.apps.santatracker.util.MeasurementManager
import com.google.android.apps.santatracker.util.SantaLog
import com.google.android.apps.santatracker.util.VectorUtil
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.storage.FirebaseStorage
import java.lang.ref.WeakReference

/** Main container for the City Quiz game.  */
class CityQuizActivity : PlayGamesActivity(), OnMapReadyCallback, GoogleMap.OnMarkerClickListener {

    private lateinit var cityQuizGame: CityQuizGame

    private lateinit var cityImageView: ImageView
    private lateinit var cityImageProgressBar: ProgressBar
    private lateinit var cloudOffImageView: ImageView
    private lateinit var cityImageAuthorTextView: TextView
    private lateinit var mapScrim: View
    private lateinit var roundCountTextView: TextView
    private lateinit var pointsTextView: TextView

    private lateinit var supportMapFragment: SupportMapFragment
    private lateinit var googleMap: GoogleMap

    private var mapLaidOut: Boolean = false
    private var mapReady: Boolean = false
    private var initialRoundLoaded: Boolean = false

    private val cityImageGlideListener = CityImageGlideListener(this)
    private val handler = Handler()

    private lateinit var analytics: FirebaseAnalytics

    override fun getLayoutId() = R.layout.activity_city_quiz

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val roundCount = FirebaseRemoteConfig.getInstance().getLong(CITY_QUIZ_ROUND_COUNT_CONFIG_KEY)
        cityQuizGame = CityQuizGame(this, roundCount.toInt())

        // [ANALYTICS]
        analytics = FirebaseAnalytics.getInstance(this)
        MeasurementManager.recordScreenView(analytics,
                getString(com.google.android.apps.santatracker.common.R.string.analytics_screen_city_quiz))

        cityImageView = findViewById(R.id.cityQuizImageView)
        cityImageProgressBar = findViewById(R.id.cityImageProgressBar)
        cloudOffImageView = findViewById(R.id.cloudOffImageView)
        cityImageAuthorTextView = findViewById(R.id.cityImageAuthorTextView)
        mapScrim = findViewById(R.id.map_scrim)
        roundCountTextView = findViewById(R.id.roundCountTextView)
        pointsTextView = findViewById(R.id.pointsTextView)

        // Clicking the "offline" image will attempt a reload
        cloudOffImageView.setOnClickListener {
            cityQuizGame.currentRound?.city?.let { city ->
                loadCityImage(city.imageName, city.imageAuthor)
            }
        }

        // Map scrim prevents clicking on map
        mapScrim.setOnClickListener {
            // No-op, eat the click
        }

        supportMapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment

        supportMapFragment.view?.doOnLayout {
            mapLaidOut = true
            loadInitialRound()
        }
        supportMapFragment.getMapAsync(this)
    }

    override fun getGameId(): String {
        return getString(com.google.android.apps.playgames.R.string.city_quiz_game_id)
    }

    override fun getGameTitle(): String {
        return getString(com.google.android.apps.santatracker.common.R.string.cityquiz)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mapReady = true
        this.googleMap = googleMap

        googleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(this,
                com.google.android.apps.santatracker.common.R.raw.map_style))
        googleMap.setOnMarkerClickListener(this)

        loadInitialRound()
    }

    /**
     * Load the initial round only if the map is ready and laid out and has not been already loaded.
     */
    private fun loadInitialRound() {
        if (mapLaidOut && mapReady && !initialRoundLoaded) {
            loadRound()
            initialRoundLoaded = true
        }
    }

    /** Load the current round of the game for user interaction.  */
    private fun loadRound() {
        if (isFinishing || supportMapFragment.view == null) {
            // The user exit the game between rounds.
            return
        }

        updateScore()

        // Get next city in game.
        val city = cityQuizGame.currentRound?.city ?: return

        // Load city image
        loadCityImage(city.imageName, city.imageAuthor)

        // Set up city markers
        googleMap.clear()

        if (BuildConfig.DEBUG) {
            SantaLog.d(TAG, "Moving to ${city.name}")
        }
        googleMap.animateCamera(
                CameraUpdateFactory.newLatLngBounds(
                        city.bounds, MapHelper.getMapPadding(supportMapFragment)))

        // Add markers and set appropriate tags.
        val locationMarker = googleMap.addMarker(
                MarkerOptions()
                        .position(city.correctLocation)
                        .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)))
        locationMarker.tag = CORRECT_MARKER

        val firsIncorrectLocationMarker = googleMap.addMarker(
                MarkerOptions()
                        .position(city.incorrectLocationOne)
                        .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)))
        firsIncorrectLocationMarker.tag = FIRST_FAKE_MARKER

        val secondIncorrectLocationMarker = googleMap.addMarker(
                MarkerOptions()
                        .position(city.incorrectLocationTwo)
                        .icon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_blue)))
        secondIncorrectLocationMarker.tag = SECOND_FAKE_MARKER
    }

    /**
     * Load the image matching the given image name into the city ImageView.
     *
     * @param imageName Name used to retrieve the image from Firebase Storage.
     * @param imageAuthor Name used to give attribution for the image.
     */
    private fun loadCityImage(imageName: String, imageAuthor: String) {
        if (isFinishing) {
            return
        }
        // Clear current image
        Glide.with(this)
                .load(com.google.android.apps.playgames.R.color.cityQuizPrimaryGreenDark)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(cityImageView)

        cityImageProgressBar.visibility = View.VISIBLE
        showImageAuthor(false)
        showOnlineUI()

        // Load new image
        val task = FirebaseStorage.getInstance().reference.child(imageName).downloadUrl
        task.addOnCompleteListener(this, LoadCityImageTaskCompleteListener(this, imageAuthor))

        // After timeout check if image URL has been retrieved, if not update UI.
        handler.postDelayed(IMAGE_LOAD_TIMEOUT_MILLIS) {
            if (!task.isComplete) {
                // Firebase was did not complete image URL retrieval.
                cityImageProgressBar.visibility = View.GONE
                showOfflineUI()
            }
        }
    }

    private class LoadCityImageTaskCompleteListener(
        activity: CityQuizActivity,
        private val imageAuthor: String
    ) : OnCompleteListener<Uri> {
        private val activityRef = WeakReference(activity)

        override fun onComplete(task: Task<Uri>) {
            val activity = activityRef.get()
            if (activity == null || activity.isFinishing) {
                return
            }
            activity.cityImageProgressBar.visibility = View.GONE

            if (task.isSuccessful) {
                val uri = task.result
                if (BuildConfig.DEBUG) {
                    SantaLog.d(TAG, "Image uri: $uri")
                }

                Glide.with(activity)
                        .load(uri)
                        .transition(DrawableTransitionOptions.withCrossFade())
                        .listener(activity.cityImageGlideListener)
                        .into(activity.cityImageView)

                activity.cityImageAuthorTextView.text = activity.resources.getString(R.string.photo_by, imageAuthor)
                activity.showImageAuthor(true)
            } else {
                if (BuildConfig.DEBUG) {
                    SantaLog.e(TAG, "Unable to get image URI from Firebase Storage", task.exception)
                }
                activity.showOfflineUI()
            }
        }
    }

    private fun showOfflineUI() {
        cloudOffImageView.visibility = View.VISIBLE
        mapScrim.visibility = View.VISIBLE
    }

    private fun showOnlineUI() {
        cloudOffImageView.visibility = View.GONE
        mapScrim.visibility = View.GONE
    }

    private fun updateScore() {
        // Update game information, round count and score.
        roundCountTextView.text = getString(
                R.string.round_count_fmt,
                cityQuizGame.currentRoundCount + 1,
                cityQuizGame.totalRoundCount)
        pointsTextView.text = getString(R.string.game_score_fmt, cityQuizGame.calculateScore())
    }

    private fun showImageAuthor(visible: Boolean) {
        if (visible) {
            cityImageAuthorTextView.visibility = TextView.VISIBLE
        } else {
            cityImageAuthorTextView.visibility = TextView.INVISIBLE
        }
    }

    override fun onMarkerClick(marker: Marker): Boolean {
        // Check if the round is already solved, if so ignore the marker click.
        if (cityQuizGame.isFinished || cityQuizGame.currentRound?.isSolved == true) {
            return true
        }

        val currentRound = cityQuizGame.currentRound ?: return false

        // Identify which marker was tapped and update the round status.
        val tag = marker.tag as Int
        currentRound.updateLocationStatus(tag, true)

        // Check if user tapped on the correct marker and if move to the next round.
        if (tag == 0) {
            marker.setIcon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_green))
            updateScore()

            // [ANALYTICS]
            val numIncorrectAttempts = 5 - currentRound.calculateRoundScore()
            MeasurementManager.recordCorrectCitySelected(
                    analytics,
                    currentRound.city.imageName,
                    numIncorrectAttempts)

            googleMap.animateCamera(
                    CameraUpdateFactory.newLatLng(currentRound.city.correctLocation),
                    object : GoogleMap.CancelableCallback {
                        override fun onFinish() {
                            showCityInfo(marker)
                        }

                        override fun onCancel() {
                            showCityInfo(marker)
                        }
                    })

            // Wait a while before moving to next round or end of game.
            handler.postDelayed(NEXT_ROUND_DELAY) {
                cityQuizGame.moveToNextRound()
                // Check if the last round has completed.
                if (!cityQuizGame.isFinished) {
                    loadRound()
                } else {
                    // or exit. For now I will clear the markers.
                    goToGameSummary()
                    pointsTextView.text = getString(
                            R.string.game_score_fmt,
                            cityQuizGame.calculateScore())
                }
            }
        } else {
            // [ANALYTICS]
            cityQuizGame.currentRound?.let { round ->
                MeasurementManager.recordIncorrectCitySelected(analytics, round.city.imageName)
            }

            marker.setIcon(VectorUtil.vectorToBitmap(this, R.drawable.ic_pin_red))
        }

        return true
    }

    private fun showCityInfo(marker: Marker) {
        marker.title = cityQuizGame.currentRound?.city?.name
        marker.showInfoWindow()
    }

    private fun goToGameSummary() {
        // Show the end-game view
        val gameView = findViewById<View>(R.id.view_end_game) as EndOfGameView
        gameView.initialize(cityQuizGame.calculateScore(),
                {
                    // Restart this activity
                    recreate()
                },
                {
                    // Back to the village
                    finish()
                })

        // Show end game view over everything
        gameView.visibility = View.VISIBLE
        gameView.z = 1000f
    }

    // Listener to handle Glide action completion.
    private class CityImageGlideListener(activity: CityQuizActivity) : RequestListener<Drawable> {
        private val weakActivity = WeakReference(activity)

        override fun onLoadFailed(
            e: GlideException?,
            model: Any?,
            target: Target<Drawable>?,
            isFirstResource: Boolean
        ): Boolean {
            // Glide failed to load image.
            if (weakActivity.get() != null) {
                SantaLog.e(TAG, "Glide unable to load city image.")
                weakActivity.get()?.showOfflineUI()
            }
            return false
        }

        override fun onResourceReady(
            resource: Drawable?,
            model: Any?,
            target: Target<Drawable>?,
            dataSource: DataSource?,
            isFirstResource: Boolean
        ): Boolean {
            if (weakActivity.get() != null) {
                // Glide loaded image, hide "cloud off" ImageView and show author TextView.
                weakActivity.get()?.showOnlineUI()
                weakActivity.get()?.showImageAuthor(true)
            }
            return false
        }
    }

    companion object {
        private const val TAG = "CityQuizActivity"
        private const val NEXT_ROUND_DELAY = 3500L
        private const val CORRECT_MARKER = 0
        private const val FIRST_FAKE_MARKER = 1
        private const val SECOND_FAKE_MARKER = 2
        private const val IMAGE_LOAD_TIMEOUT_MILLIS = 5000L // 5 seconds
        private const val CITY_QUIZ_ROUND_COUNT_CONFIG_KEY = "CityQuizRoundCount"
    }
}
